### PR TITLE
Build devnet image for testing purposes

### LINKED ISSUE #187: Add build stage for devnet creation
## 📚 Context

Create a new build stage for devnet by leveraging #149 to bring an instance of anvil up for deploying a CartesiDApp.

## ✔️ Solution

Build a devnet image based on a machine snapshot (see #173).

## 📈 Subtasks

- [x] Add new buildx target for devnet
- [x] Remove deployer from compose
- [x] replace `sunodo/devnet` with `node/devnet` in compose

### PR DESCRIPTION
**Pull request by marcelstanley**
We're not deploying the contracts in a deterministic fashion yet. That's the reason for [da0c33e](https://github.com/cartesi/rollups-node/pull/204/commits/da0c33e877b82b7139244034057460557018708b), which is expected to be reverted soon as part of #229.

---
### Testing

Please make sure to build the image:

```sh
cd build
docker buildx bake rollups-devnet --load --no-cache
```

Also, check if the state file has the expected size as follows:

```sh
cd /tmp
id=$(docker create cartesi/rollups-devnet:devel)
docker cp $id:/usr/share/devnet/anvil_state.json .
docker rm -v $id
ls -l anvil_state.json
cat anvil_state.json | jq >> /dev/null
```

Expected output (file size is `82622 bytes` and there are no errors):

```sh
Successfully copied 84.5kB to /tmp/.
7e918c31f5177610196810c7eedd61212670840f5ed05e96fd4b348294645cac
-rw-r--r-- 1 username username 82622 Dec 15 16:12 anvil_state.json
```

---

Closes #187

### CODE REVIEW

File: build/node-compose.yml [hunk 1]
```diff
@@ -29,11 +27,12 @@ services:
...
-      CARTESI_CONTRACTS_DAPP_ADDRESS: "0x70ac08179605AF2D9e75782b8DEcDD3c22aA4D0C"
+      CARTESI_BLOCKCHAIN_READ_DEPTH: "1"
+      CARTESI_CONTRACTS_DAPP_ADDRESS: "0x180763470853cAF642Df79a908F9282c61692A45"
       CARTESI_CONTRACTS_DAPP_DEPLOYMENT_BLOCK_NUMBER: "1"
-      CARTESI_CONTRACTS_HISTORY_ADDRESS: "0x4FF8BD9122b7D91d56Dd5c88FE6891Fb3c0b5281"
-      CARTESI_CONTRACTS_AUTHORITY_ADDRESS: "0x5050F233F2312B1636eb7CF6c7876D9cC6ac4785"
-      CARTESI_CONTRACTS_INPUT_BOX_ADDRESS: "0x59b22D57D4f067708AB0c00552767405926dc768"
+      CARTESI_CONTRACTS_HISTORY_ADDRESS: "0xbCc4067dfdeE8274E531951f0968EC895f397A44"
+      CARTESI_CONTRACTS_AUTHORITY_ADDRESS: "0xb090149a3cA43000681B74875500854B54355496"
+      CARTESI_CONTRACTS_INPUT_BOX_ADDRESS: "0xCf7Ed3AccA5a467e9e704C703E8D87F634fB0Fc9"
```
Comment (tuler) [inline build/node-compose.yml:35]: why doesn't it have the same addresses of the testnets and mainnets?

File: pkg/addresses/addresses.go [hunk 2]
```diff
@@ -33,15 +33,15 @@ type Book struct {
...
-		CartesiDAppFactory:  common.HexToAddress("0x7122cd1221C20892234186facfE8615e6743Ab02"),
-		DAppAddressRelay:    common.HexToAddress("0xF5DE34d6BbC0446E2a45719E718efEbaaE179daE"),
-		ERC1155BatchPortal:  common.HexToAddress("0xedB53860A6B52bbb7561Ad596416ee9965B055Aa"),
-		ERC1155SinglePortal: common.HexToAddress("0x7CFB0193Ca87eB6e48056885E026552c3A941FC4"),
-		ERC20Portal:         common.HexToAddress("0x9C21AEb2093C32DDbC53eEF24B873BDCd1aDa1DB"),
-		ERC721Portal:        common.HexToAddress("0x237F8DD094C0e47f4236f12b4Fa01d6Dae89fb87"),
-		EtherPortal:         common.HexToAddress("0xFfdbe43d4c855BF7e0f105c400A50857f53AB044"),
-		InputBox:            common.HexToAddress("0x59b22D57D4f067708AB0c00552767405926dc768"),
-		CartesiDApp:         common.HexToAddress("0x70ac08179605AF2D9e75782b8DEcDD3c22aA4D0C"),
+		CartesiDAppFactory:  common.HexToAddress("0x610178dA211FEF7D417bC0e6FeD39F05609AD788"),
```
Comment (tuler) [inline pkg/addresses/addresses.go:36]: does this need to be hard-coded?

File: build/node-compose.yml (hunk 1 above)
Comment (marcelstanley) [reply build/node-compose.yml:35]: See #229

File: pkg/addresses/addresses.go (hunk 2 above)
Comment (marcelstanley) [reply pkg/addresses/addresses.go:36]: AFAIK, this is just to simplify our internal test/dev tool. It was a development decision. We may change it to no rely on hardcoded values in the future if needed be.

File: build/Dockerfile [hunk 3]
```diff
@@ -54,6 +56,54 @@ RUN ln -s ${MACHINE_SNAPSHOT_PATH}/0_0 ${MACHINE_SNAPSHOT_PATH}/latest
...
+    ca-certificates \
+    curl \
+    git \
+    jq \
+    wget \
+    xxd
+
+# Install foundry
+RUN curl -L https://foundry.paradigm.xyz | bash
+SHELL ["/bin/bash", "-c"]
```
Comment (gligneul) [inline build/Dockerfile:76]: Why do we need to change the shell?

File: build/node-compose.yml [hunk 4]
```diff
@@ -29,11 +27,12 @@ services:
       CARTESI_BLOCKCHAIN_IS_LEGACY: "false"
       CARTESI_BLOCKCHAIN_GENESIS_BLOCK: "1"
       CARTESI_BLOCKCHAIN_FINALITY_OFFSET: "1"
-      CARTESI_CONTRACTS_DAPP_ADDRESS: "0x70ac08179605AF2D9e75782b8DEcDD3c22aA4D0C"
+      CARTESI_BLOCKCHAIN_READ_DEPTH: "1"
```
Comment (gligneul) [inline build/node-compose.yml:30]: This variable doesn't exist; see https://github.com/cartesi/rollups-node/blob/main/docs/config.md

File: pkg/addresses/addresses.go (hunk 2 above)
Comment (gligneul) [reply pkg/addresses/addresses.go:36]: It is hardcoded to simplify our lives, but ideally, this file should be generated automatically with `go generate`. Marcel, could you create an issue to tackle that?

**Review by gligneul — CHANGES_REQUESTED** [review]

File: pkg/addresses/addresses.go (hunk 2 above)
Comment (marcelstanley) [reply pkg/addresses/addresses.go:36]: Issue #233 created

File: build/Dockerfile (hunk 3 above)
Comment (marcelstanley) [reply build/Dockerfile:74]: It's a workaround for https://github.com/foundry-rs/foundry/issues/2476. Another alternative would be calling `foundryup` like this:
```docker
RUN ~/.foundry/bin/foundryup
```

File: build/node-compose.yml (hunk 4 above)
Comment (marcelstanley) [reply build/node-compose.yml]: I don't recall where it comes from. Removed at 9e6a40852196dad533525be4c928c4b542c650cb

**Review by gligneul — APPROVED** [review]

**Pull request merged**